### PR TITLE
zig_0_13: init

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

New zig version released.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/compilers/zig/0.13/default.nix [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    (lib.cmakeFeature "ZIG_TARGET_MCPU" "baseline")
+    # always link against static build of LLVM
+    (lib.cmakeBool "ZIG_STATIC_LLVM" true)
+  ];
+
+  env.ZIG_GLOBAL_CACHE_DIR = "$TMPDIR/zig-cache";
+
+  doInstallCheck = true;
+
+  strictDeps = true;
```
[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix]: Fixed. I made sure to disable on Darwin and when LLVM is in use.

[the remark this answers is not in the document]
Comment (paparodeo) [reply pkgs/development/compilers/zig/0.13/default.nix]:
```nix
strictDeps = !stdenv.isDarwin || !(stdenv.hostPlatform.useLLVM or false);
```
will be true when `stdenv = clangStdenv` and thus will fail. without thinking too hard i think `strictDeps = !stdenv.cc.isClang;` is sufficient but maybe i'm missing something.

[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix]: Alright, yeah. That does work.

**Comment by AndersonTorres** [comment]
This is not an update, but a new package.
Please reword the commit as `zig_0_13: init`

File: pkgs/development/compilers/zig/0.13/default.nix [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    (lib.cmakeBool "ZIG_STATIC_LLVM" true)
+  ];
+
+  env.ZIG_GLOBAL_CACHE_DIR = "$TMPDIR/zig-cache";
+
+  doInstallCheck = true;
+
+  # strictDeps breaks zig when clang is being used.
+  # https://github.com/NixOS/nixpkgs/issues/317055#issuecomment-2148438395
+  strictDeps = !stdenv.cc.isClang;
```
Comment (AndersonTorres) [inline pkgs/development/compilers/zig/0.13/default.nix:61]: Being honest, I do not care about breakages in Darwin.
Nonetheless, setting strictDeps as false - even in special cases - is a downgrade.

Since a working package is better than no package, I will not block it, at least for now.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -25038,7 +25038,11 @@ with pkgs;
   zig_0_12 = darwin.apple_sdk_11_0.callPackage ../development/compilers/zig/0.12 {
     llvmPackages = llvmPackages_17;
   };
-  zig = zig_0_12;
+  # requires a newer Apple SDK
+  zig_0_13 = darwin.apple_sdk_11_0.callPackage ../development/compilers/zig/0.13 {
+    llvmPackages = llvmPackages_18;
+  };
+  zig = zig_0_13;
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:25045]: This "update" should be in a separate commit, namely,  `zig: set as zig 0.13`

Comment (RossComputerGuy) [reply pkgs/top-level/all-packages.nix:25045]: Done

File: pkgs/development/compilers/zig/0.13/default.nix (hunk 2 above)
Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix:61]: True though I don't think of `strictDeps` being `false` is a downgrade when it causes a breakage in certain scenarios. We should look into fixing this properly but a different PR can do that.

File: pkgs/development/compilers/zig/0.13/default.nix [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  postPatch = ''
+    substituteInPlace lib/std/zig/system.zig \
+      --replace "/usr/bin/env" "${lib.getExe' coreutils "env"}"
+  '';
+
+  postBuild = ''
+    mkdir -p codedir
+
+    for file in ../doc/langref/*.zig; do
+      stage3/bin/zig run ../tools/doctest.zig -- --zig $PWD/stage3/bin/zig -i "$file" -o codedir/$(basename "$file").out
```
Comment (RossComputerGuy) [inline pkgs/development/compilers/zig/0.13/default.nix:74]: Builds appear to be failing here since `-i` doesn't seem to be expecting spaces in the filenames in `doc/langref/`.

Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix]: It seems `-o` might be the problem.

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/development/compilers/zig/0.13/default.nix (hunk 4 above)
Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix]: Fixed

File: pkgs/development/compilers/zig/0.13/default.nix [hunk 5]
```diff
@@ -73,6 +73,7 @@ stdenv.mkDerivation (finalAttrs: {
 
     for file in ../doc/langref/*.zig; do
       stage3/bin/zig run ../tools/doctest.zig -- --zig $PWD/stage3/bin/zig --cache-root cache-root -i "$file" -o "codedir/$(basename "$file").out"
+      cp "$file" codedir/
```
Comment (AndersonTorres) [inline pkgs/development/compilers/zig/0.13/default.nix:76]: 1. Why this copy? Can't this file be generated already inside codedir in the previous command?
2. this should be squashed with the other commit.

Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix:76]: 1. Not sure, I've been trying to get Zig working with little success
2. I thought I squashed it but must've forgot

Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix]: Looking through the errors on Ofborg, it seems I mixed up the docgen needing a file called `hello.out` but I thought it was `hello.zig`. I made an adjustment with how `doctest` is called so this should work now. I also squashed the commit.

Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix]: Got as far as `./doc/langref.html.in:2878:14: error: unable to open 'handle_error_with_catch_block.zig.out': FileNotFound`. I think this next build I try might work.

File: pkgs/development/compilers/zig/0.13/default.nix [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+  postBuild = ''
+    mkdir -p codedir cache-root/tmp
+    ln -s ../../../doc cache-root/tmp/doc
+
+    for file in ../doc/langref/*.zig; do
+      stage3/bin/zig run ../tools/doctest.zig -- --zig $PWD/stage3/bin/zig --cache-root cache-root -i "$file" -o "codedir/$(basename "$file" | rev | sed 's/giz./tuo./' | rev)"
+    done
+
+    stage3/bin/zig run ../tools/docgen.zig -- --code-dir codedir ../doc/langref.html.in langref.html
```
Comment (andrewrk) [inline pkgs/development/compilers/zig/0.13/default.nix:80]: why have you done this? `zig build` already builds the langref.

it's not wise to duplicate the upstream's build logic in shell scripts in the nix package.

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/zig/0.13/default.nix (hunk 6 above)
Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix:80]: Because CMake doesn't do it.

Comment (andrewrk) [reply pkgs/development/compilers/zig/0.13/default.nix:80]: `stage3/bin/zig build langref`

Comment (RossComputerGuy) [reply pkgs/development/compilers/zig/0.13/default.nix]: Alright, didn't know that was a step in Zig's `build.zig`.

File: pkgs/development/compilers/zig/0.13/hook.nix [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    # This zig_default_flags below is meant to avoid CPU feature impurity in
+    # Nixpkgs. However, this flagset is "unstable": it is specifically meant to
+    # be controlled by the upstream development team - being up to that team
+    # exposing or not that flags to the outside (especially the package manager
+    # teams).
+
+    # Because of this hurdle, @andrewrk from Zig Software Foundation proposed
+    # some solutions for this issue. Hopefully they will be implemented in
+    # future releases of Zig. When this happens, this flagset should be
+    # revisited accordingly.
```
Comment (andrewrk) [inline pkgs/development/compilers/zig/0.13/hook.nix:22]: This has happened, they were implemented

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Review by andrewrk — CHANGES_REQUESTED** [review]
It looks like you uncritically deleted the text

> This zig_default_flags below is meant to avoid CPU feature impurity in
> Nixpkgs. However, this flagset is "unstable": it is specifically meant to
> be controlled by the upstream development team - being up to that team
> exposing or not that flags to the outside (especially the package manager
> teams).

without stopping to think about the implications of what you are doing and what that means for this package.

Frankly, this is sloppy work. I think this pull request should be reviewed carefully before being merged.

**Comment by RossComputerGuy** [comment]
> It looks like you uncritically deleted the text

Didn't realize that was selected when I deleted the text you mentioned to delete. I readded it. The build I was running before the one calling `zig build langref` finally finished with no failure.

> Frankly, this is sloppy work. I think this pull request should be reviewed carefully before being merged.

Sorry, it is hard to test this on NixOS on Apple Silicon so I have to push and then build remotely. I also wasn't aware of the changes with the documentation generation which frankly was the most difficult part of this. It is also harder to rebuild when the build times are an hour long, used to be 10 minutes on the same system.

**Comment by RossComputerGuy** [comment]
Zig finally builds

**Comment by RossComputerGuy** [comment]
Forgot to remove `util-linux` after switching to calling `zig` for the langref. It has been removed. Ofborg built everything except 1 instance of zig, both linux pass and 1 darwin passed so this should be good to go now.

**Comment by tomberek** [comment]
Result of `nixpkgs-review pr 317861` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>zig (zig_0_13)</li>
    <li>zig.doc (zig_0_13.doc)</li>
  </ul>
</details>
